In [2]:
import pyxdf
import numpy as np
import matplotlib.pyplot as plt
import json

In [3]:
import os

# Construct paths to local data files
# Navigate up one level from test-reve/ to workspace root, then into data/
data_dir = os.path.join(os.path.dirname(os.getcwd()), 'data')
xdf_file_1 = os.path.join(data_dir, 'sub-PD089', 'ses-S001', 'eeg', 'sub-PD089_ses-S001_task-demo_workload_run-001_eeg.xdf')
xdf_file_2 = os.path.join(data_dir, 'sub-PD094', 'ses-S001', 'eeg', 'sub-PD094_ses-S001_task-demo_workload_run-001_eeg.xdf')

streams1, header1 = pyxdf.load_xdf(xdf_file_1)
streams2, header2 = pyxdf.load_xdf(xdf_file_2)

Each `stream` object in the list `streamsX` is a dictionary, and typically contains the following important keys:

*   **`'info'`**: This is a dictionary holding metadata about the stream. It contains information such as:
    *   `'name'`: The name of the stream (e.g., 'actiCHampMarkers', 'actiCHamp-24020239', 'ZLT-logging').
    *   `'type'`: The type of data in the stream (e.g., 'Markers', 'EEG', 'misc', 'Logging').
    *   `'channel_count'`: The number of channels if it's a multi-channel stream.
    *   `'nominal_srate'`: The nominal sampling rate (samples per second).
    *   `'source_id'`: A unique identifier for the hardware source.
    *   `'created_at'`: Timestamp of when the stream was created.
    *   `'desc'`: A dictionary that can contain more detailed, custom descriptions or metadata for the stream, often including channel labels.

*   **`'time_series'`**: This is the actual data recorded in the stream. Its format depends on the stream type:
    *   For EEG or other numeric data, it's typically a NumPy array where rows are samples and columns are channels.
    *   For Marker streams, it's often a list of strings, where each string represents a marker event.

*   **`'time_stamps'`**: This is a NumPy array of timestamps, synchronized with the `'time_series'`. Each timestamp corresponds to a sample or a marker event.

In [4]:
print(f"\nNumber of streams in file 1: {len(streams1)}")
print(f"\nNumber of streams in file 2: {len(streams2)}")


Number of streams in file 1: 7

Number of streams in file 2: 7


In [5]:
for i, stream in enumerate(streams1):
    print(f"\n=== Stream {i+1} ===")

    # Basic metadata
    name = stream['info']['name'][0]
    stype = stream['info']['type'][0]
    print("Name:", name)
    print("Type:", stype)

    # Time-series structure
    ts = stream['time_series']
    n_samples = len(ts)
    print("Samples:", n_samples)

    # Detect channels robustly
    if n_samples == 0:
        print("Channels: 0 (empty stream)")
    else:
        first_sample = ts[0]
        if isinstance(first_sample, (list, tuple)):
            print("Channels:", len(first_sample))
        else:
            print("Channels: 1")

        # Show first sample safely
        print("First sample:", first_sample)

    # Timestamps info
    print("Timestamps:", len(stream['time_stamps']))


=== Stream 1 ===
Name: actiCHampMarkers-24020239
Type: Markers
Samples: 0
Channels: 0 (empty stream)
Timestamps: 0

=== Stream 2 ===
Name: captrak
Type: misc
Samples: 0
Channels: 0 (empty stream)
Timestamps: 0

=== Stream 3 ===
Name: ZLT-logging
Type: Logging
Samples: 76
Channels: 1
First sample: ['user id: ']
Timestamps: 76

=== Stream 4 ===
Name: NAFAS-markers
Type: Notes
Samples: 0
Channels: 0 (empty stream)
Timestamps: 0

=== Stream 5 ===
Name: actiCHamp-24020239
Type: EEG
Samples: 912513
Channels: 1
First sample: [ 1.17197266e+03  1.04800293e+04  1.57700195e+03  1.88955078e+03
  3.21386719e+03  6.88085938e+02 -4.50341797e+02 -2.55844727e+03
 -9.47119141e+02 -2.08134766e+03 -4.02783203e+03 -8.40083008e+03
 -5.43095703e+03 -2.24702148e+03 -3.89965820e+03  6.46518555e+03
 -1.00293457e+04 -7.48842773e+03 -9.07744141e+03 -7.97514648e+03
  2.44042969e+03 -1.82734375e+03 -6.83071289e+03 -5.27543945e+03
 -1.19762207e+04 -1.45747070e+04 -1.44814941e+04 -7.19985352e+03
 -5.84189453e+03 -8.

In [6]:
for i, stream in enumerate(streams2):
    print(f"\n=== Stream {i+1} ===")

    # Basic metadata
    name = stream['info']['name'][0]
    stype = stream['info']['type'][0]
    print("Name:", name)
    print("Type:", stype)

    # Time-series structure
    ts = stream['time_series']
    n_samples = len(ts)
    print("Samples:", n_samples)

    # Detect channels robustly
    if n_samples == 0:
        print("Channels: 0 (empty stream)")
    else:
        first_sample = ts[0]
        if isinstance(first_sample, (list, tuple)):
            print("Channels:", len(first_sample))
        else:
            print("Channels: 1")

        # Show first sample safely
        print("First sample:", first_sample)

    # Timestamps info
    print("Timestamps:", len(stream['time_stamps']))


=== Stream 1 ===
Name: impedance
Type: misc
Samples: 0
Channels: 0 (empty stream)
Timestamps: 0

=== Stream 2 ===
Name: ZLT-logging
Type: Logging
Samples: 78
Channels: 1
First sample: ['user id: ']
Timestamps: 78

=== Stream 3 ===
Name: actiCHamp-24020239
Type: EEG
Samples: 885518
Channels: 1
First sample: [ 8.4509277e+03  6.1495117e+03  1.1002637e+04  1.1380566e+04
  8.2115234e+03  7.4686523e+03  7.5339355e+03  6.6419434e+03
  7.8516113e+03  4.1802734e+03  5.3208008e+03  6.4093262e+03
  1.4285645e+04  6.7459473e+03  8.4293457e+03  9.0986328e+03
 -8.0756836e+03 -1.3195850e+04  9.1918945e+02  2.2597656e+02
 -5.4259766e+03 -1.0295801e+04 -9.7043457e+03 -9.9939453e+03
  1.1317969e+04  2.0085938e+03 -1.2401074e+04 -9.9903809e+03
 -9.9184570e+03 -9.9528809e+03 -1.0376123e+04 -1.1617773e+04
  9.6196777e+03  7.1156250e+03  2.9600586e+03  5.2782227e+03
  9.4030762e+03  1.0922607e+04  5.2956055e+03  4.0936035e+03
  1.9890967e+04  2.2353027e+03  1.1397363e+04  2.2529834e+04
  3.1529785e+03  1.0

# Summary of XDF Data Streams:

* **Stream 1: 'actiCHampMarkers-24020239' (Type: Markers)**
    * **Inference:** This stream is **empty** (0 samples) and contains no usable marker data.

* **Stream 2: 'captrak' (Type: misc)**
    * **Inference:** This stream is **empty** (0 samples). It was likely intended for motion capture data but is unused and can be ignored.

* **Stream 3: 'ZLT-logging' (Type: Logging)**
    * **Inference:** Contains **session metadata** (76 samples), such as `['user id: ']`. Useful for high-level session tracking but not for primary event marking.

* **Stream 4: 'NAFAS-markers' (Type: Notes)**
    * **Inference:** This stream is **empty** (0 samples) and can be ignored.

* **Stream 5: 'actiCHamp-24020239' (Type: EEG)**
    * **Inference:** This is the **Primary EEG Data Stream** (912,513 samples). It holds the raw continuous brain activity. The reported channel count of '1' is misleading in the summary due to parsing limitations; the subsequent check confirmed **99 channels** in this stream.

* **Stream 6: 'ZLT-markers' (Type: Markers)**
    * **Inference:** This is the **Primary Event Marker Stream** (2,800 samples). It contains the experimental events (e.g., `['Recording/Start']`) needed to identify epochs and link the EEG data to stimuli/conditions.

* **Stream 7: 'impedance' (Type: misc)**
    * **Inference:** This stream is **empty** (0 samples). It was likely intended for impedance logging but recorded no samples and can be ignored.

In [7]:
eeg_stream = streams1[4]
marker_stream = streams1[5]

eeg = np.array(eeg_stream['time_series'])
t = np.array(eeg_stream['time_stamps'])
marker_data = marker_stream['time_series']
marker_timestamps = marker_stream['time_stamps']


In [8]:
# Define the custom serialization function
def numpy_to_python(obj):
    """Custom JSON serializer for NumPy data types"""
    if isinstance(obj, (np.int_, np.intc, np.intp, np.int8,
                        np.int16, np.int32, np.int64, np.uint8,
                        np.uint16, np.uint32, np.uint64)):
        return int(obj)
    elif isinstance(obj, (np.float_, np.float16, np.float32, np.float64)):
        return float(obj)
    elif isinstance(obj, (np.ndarray,)):
        return obj.tolist()
    raise TypeError(f"Object of type {obj.__class__.__name__} is not JSON serializable")

# Get the EEG info dictionary (Stream 5 is at index 4)
eeg_info = eeg_stream['info']

print("--- Full EEG Stream Info ---")
# Apply the custom serializer using the 'default' argument
# The user will need to run this command in their environment
print(json.dumps(eeg_info, indent=4, default=numpy_to_python))

sfreq = float(eeg_info['nominal_srate'][0])
duration_sec = eeg.shape[0] / sfreq

print(f"--- 1. Time Domain Analysis ---")
print(f"EEG Sampling Rate (Hz): {sfreq}")
print(f"Total Recording Duration: {duration_sec:.2f} seconds ({duration_sec / 60:.2f} minutes)")
print(f"Expected Samples (Duration * Fs): {duration_sec * sfreq:.0f}")

--- Full EEG Stream Info ---
{
    "name": [
        "actiCHamp-24020239"
    ],
    "type": [
        "EEG"
    ],
    "channel_count": [
        "99"
    ],
    "channel_format": [
        "float32"
    ],
    "source_id": [
        "24020239"
    ],
    "nominal_srate": [
        "500.0000000000000"
    ],
    "version": [
        "1.100000000000000"
    ],
    "created_at": [
        "500330.3321765000"
    ],
    "uid": [
        "6e443888-d495-4fea-af0c-c0328853bfc0"
    ],
    "session_id": [
        "default"
    ],
    "hostname": [
        "Participant"
    ],
    "v4address": [
        null
    ],
    "v4data_port": [
        "16574"
    ],
    "v4service_port": [
        "16586"
    ],
    "v6address": [
        null
    ],
    "v6data_port": [
        "16574"
    ],
    "v6service_port": [
        "16586"
    ],
    "desc": [
        {
            "channels": [
                {
                    "channel": [
                        {
                            "label":

In [9]:
# 2. Identify Channel Labels (Spatial Domain)
# The channel labels are nested deeply within the 'info' structure.
# Use a list comprehension to extract them.
channel_labels = [ch['label'][0] for ch in eeg_info['desc'][0]['channels'][0]['channel']]

print(f"--- 2. Channel Label Analysis ---")
print(f"Number of channel labels found: {len(channel_labels)}")
print("First 5 channel labels:", channel_labels[:5])
print("Last 5 channel labels:", channel_labels[-5:])  # What are AUX1 and AUX2?

--- 2. Channel Label Analysis ---
Number of channel labels found: 99
First 5 channel labels: ['1', '2', '3', '4', '5']
Last 5 channel labels: ['95', '96', 'AUX_1', 'AUX_2', 'Markers']


In [10]:
# 3. Explore Marker Stream Content
# Use a set to quickly find all unique marker strings.
unique_markers = set([item[0] for item in marker_data])

print(f"--- 3. Marker Content Analysis ---")
print(f"Number of unique markers: {len(unique_markers)}")
print("Unique Marker Labels:", unique_markers)

--- 3. Marker Content Analysis ---
Number of unique markers: 263
Unique Marker Labels: {'offset no task - sparkles - breathing with counting(Def/PD, Onset)', 'no task - sparkles - breathing with counting: 10(Def/PD, Onset)', 'no task - no sparkles - recurrent addition: 6(Def/PD, Onset)', 'no task - sparkles - recurrent addition(Def/PD, Onset)', 'no task - no sparkles - breathing with counting: 10(Def/PD, Onset)', 'onset no task - sparkles - fixation cross(Def/PD, Onset)', 'task - sparkles - high workload: 5(Def/PD, Onset)', 'task - no sparkles - medium workload: 1(Def/PD, Onset)', 'onset no task - no sparkles - recurrent addition(Def/PD, Onset)', '(Organizational-property/Practice-trial/During): onset task - sparkles - original workload(Def/PD, Onset)', 'no task - sparkles - fixation cross: 5(Def/PD, Onset)', 'task - sparkles - original workload: 10(Def/PD, Onset)', '(Organizational-property/Practice-trial/During): task - sparkles - original workload: 7(Def/PD, Onset)', 'no task - no s

In [11]:
# 4. Time Synchronization Check
# Check if the marker timestamps align with the EEG timestamps.
print(f"--- 4. Time Synchronization Check ---")
first_eeg_ts = t[0]
first_marker_ts = marker_timestamps[0]
print(f"EEG recording start time (LSL): {first_eeg_ts}")
print(f"Marker stream first timestamp (LSL): {first_marker_ts}")
print(f"Time difference between stream starts: {first_eeg_ts - first_marker_ts:.4f} seconds")

--- 4. Time Synchronization Check ---
EEG recording start time (LSL): 494561.7709630675
Marker stream first timestamp (LSL): 494565.1925672509
Time difference between stream starts: -3.4216 seconds


In [12]:
logging_stream = streams1[2]  # Stream 3 is at index 2
log_data = logging_stream['time_series']
log_timestamps = logging_stream['time_stamps']

print(f"Total number of log entries: {len(log_data)}")
print("\n--- First 5 Log Entries and Timestamps ---")

# Iterate over the first 5 entries
for i in range(min(20, len(log_data))):
    # Safely extract the log entry (assuming it's a list containing a single string)
    log_entry = log_data[i][0] if isinstance(log_data[i], (list, tuple)) and len(log_data[i]) > 0 else log_data[i]
    timestamp = log_timestamps[i]
    print(f"Timestamp: {timestamp:.4f} | Content: {log_entry}")

print("\n--- Last 5 Log Entries and Timestamps ---")

# Print the last 5 logs if there are more than 5
if len(log_data) > 5:
    for i in range(len(log_data) - 5, len(log_data)):
        # Safely extract the log entry
        log_entry = log_data[i][0] if isinstance(log_data[i], (list, tuple)) and len(log_data[i]) > 0 else log_data[i]
        timestamp = log_timestamps[i]
        print(f"Timestamp: {timestamp:.4f} | Content: {log_entry}")

Total number of log entries: 76

--- First 5 Log Entries and Timestamps ---
Timestamp: 494565.1922 | Content: user id: 
Timestamp: 494565.1926 | Content: Recording/Start, Description/DemoWorkload
Timestamp: 494632.4333 | Content: (Organizational-property/Practice-trial/During): (Participant-response, Description/correct user-probe)
Timestamp: 494778.7744 | Content: Time-block, Item-count/01
Timestamp: 494811.6407 | Content: (Participant-response, Description/incorrect user-probe)
Timestamp: 494836.6649 | Content: (Participant-response, Description/no user-probe)
Timestamp: 494861.6900 | Content: (Participant-response, Description/no user-probe)
Timestamp: 494886.7145 | Content: (Participant-response, Description/no user-probe)
Timestamp: 494911.7392 | Content: (Participant-response, Description/no user-probe)
Timestamp: 494915.0389 | Content: Time-block, Item-count/02
Timestamp: 494940.0635 | Content: (Participant-response, Description/no user-probe)
Timestamp: 494969.0477 | Content: (

In [13]:
# The log stream provided two key time points: Recording/Start and Recording/End.
# We can use these to calculate the log stream duration and compare it to the EEG duration.

print("\n--- Next Analysis Step: Log Stream Duration ---")

# The first and last timestamps provided in the output
log_start_ts = 494565.1926
log_end_ts = 496382.9123

log_duration = log_end_ts - log_start_ts
print(f"Total time logged (based on start/end events): {log_duration:.4f} seconds ({log_duration / 60:.2f} minutes)")

# You should compare this duration to the EEG duration calculated in the previous step
# to ensure the EEG data completely spans the logged events.

# Next, we must focus on *extracting* the markers that define your classification task.


--- Next Analysis Step: Log Stream Duration ---
Total time logged (based on start/end events): 1817.7197 seconds (30.30 minutes)


In [14]:
from collections import Counter

marker_stream = streams1[5]  # Stream 6 (ZLT-markers) is at index 5
marker_data = marker_stream['time_series']

# Convert the nested list structure (e.g., [['Start'], ['End']]) into a flat list of strings
# We use a try-except block to handle cases where marker_data might be a NumPy array.
try:
    # Flatten the data: marker_data is usually a list of single-item lists
    flat_markers = [item[0] for item in marker_data]
except:
    # Fallback assumption if it's a 1D NumPy array or already flat
    flat_markers = marker_data

# Count the frequency of each unique marker
marker_counts = Counter(flat_markers)

print(f"Total markers in ZLT-markers stream: {len(flat_markers)}")
print("\n--- Marker Frequency Analysis (Unique Markers and Counts) ---")

# Print the results sorted by frequency (most common first)
for marker, count in marker_counts.most_common():
    print(f"Marker: {marker:<30} | Count: {count}")

Total markers in ZLT-markers stream: 2800

--- Marker Frequency Analysis (Unique Markers and Counts) ---
Marker: no task - sparkles - fixation cross(Def/PD, Onset) | Count: 100
Marker: task - no sparkles - original workload(Def/PD, Onset) | Count: 100
Marker: no task - no sparkles - recurrent addition(Def/PD, Onset) | Count: 100
Marker: task - sparkles - low workload(Def/PD, Onset) | Count: 100
Marker: task - no sparkles - medium workload(Def/PD, Onset) | Count: 90
Marker: no task - sparkles - breathing without counting(Def/PD, Onset) | Count: 90
Marker: no task - no sparkles - breathing with counting(Def/PD, Onset) | Count: 80
Marker: task - sparkles - high workload(Def/PD, Onset) | Count: 80
Marker: no task - sparkles - breathing with counting(Def/PD, Onset) | Count: 70
Marker: task - no sparkles - high workload(Def/PD, Onset) | Count: 70
Marker: Experiment-procedure/Participant-probe(Def/PD, Onset) | Count: 60
Marker: no task - no sparkles - breathing without counting(Def/PD, Onset)

In [15]:
# for stream in streams1:
#     y = stream["time_series"]

#     if isinstance(y, list):
#         # list of strings, draw one vertical line for each marker
#         for timestamp, marker in zip(stream["time_stamps"], y):
#             plt.axvline(x=timestamp)
#             print(f'Marker "{marker[0]}" @ {timestamp:.2f}s')
#     elif isinstance(y, np.ndarray):
#         # numeric data, draw as lines
#         plt.plot(stream["time_stamps"], y)
#     else:
#         raise RuntimeError("Unknown stream format")

# plt.show()